analyses/org.eclipse.tracecompass.incubator.inandout.core/src/org/eclipse/tracecompass/incubator/internal/inandout/core/analysis/InAndOutAnalysisModule.java

```diff
@@ -50,23 +58,79 @@ public class InAndOutAnalysisModule extends InstrumentedCallStackAnalysis {
      */
     public static final String JSON = ".config.json"; //$NON-NLS-1$
 
+    private @Nullable ITmfConfiguration fConfiguartion;
+    private @Nullable List<@NonNull SegmentSpecifier> fSpecifiers;
+
+    @Override
+    public @NonNull String getName() {
+        ITmfConfiguration config = fConfiguartion;
+        if (config != null) {
+            return config.getName();
+        }
+        return super.getName();
+    }
+
+    @Override
+    public @NonNull String getHelpText() {
+        ITmfConfiguration config = fConfiguartion;
+        if (config != null) {
+            return config.getDescription();
+        }
+        return super.getHelpText();
+    }
+
+    @Override
+    public boolean setConfiguration(ITmfConfiguration configuration) {
+        if (configuration == null) {
+            return false;
+        }
+        try {
+            fConfiguartion = configuration;
+            String json = configuration.getJsonParameters();
+            if (json.equals("{}")) { //$NON-NLS-1$
+                throw new TmfConfigurationException("JSON parameter is not deserialize correctly"); //$NON-NLS-1$
+            }
+            SegmentSpecifierConfiguration config = SegmentSpecifierConfiguration.fromJsonString(json);
+            fSpecifiers = config.getSpecifiers();
```

Comment:
For this example, it looks straight forward. In general, extension have to deserialize the JSON. If we pass an already deserialized JSON in map<String, Object> then extension needs to traverse the map (which could be map of maps>, instead of using deserialization using POJOs. This is a design choice. Not sure which one we should choose. 

If we go with the Map approach it would be consistent with the rest of cases in the server. Validation would have to be manual in the extension and we cannot use a validation library for schema validation unless the map is serialized back to JSON and fed to the validator.

If we go with the string, then many existing code needs to be deprecated to remove 2 ways of configuring. With a string, the extension can handle the JSON string as required, POJO, parameter map. But this would be different way of working and serialization/deserialization of json.

I need to think about it a bit more.

Reply:
Right now we have DataProviderService deserializing the parameters into a JsonNode (in ConfigurationQueryParameters class). Then it converts the JsonNode to string in ITmfDataProviderSource.createDataProviderDescriptors(). Later the analysis module deserializes the parameters string into a SegmentSpecifierConfiguration.

What if we had just left the DataProviderService use the default to deserialize the parameters into a Map. Then nothing would prevent the extension (the analysis module) to still be able to deserialize the parameters into a SegmentSpecifierConfiguration. It just has to get the json string from this at line 89: 

            String json = new Gson().toJson(configuration.getParameters());

So some extensions can optionally validate and deserialize the json string, when it's a complex object, and others can just use the map of parameters when it's a simple few values.

Reply:
> Right now we have DataProviderService deserializing the parameters into a JsonNode (in ConfigurationQueryParameters class). Then it converts the JsonNode to string in ITmfDataProviderSource.createDataProviderDescriptors(). Later the analysis module deserializes the parameters string into a SegmentSpecifierConfiguration.
> 
> What if we had just left the DataProviderService use the default to deserialize the parameters into a Map. Then nothing would prevent the extension (the analysis module) to still be able to deserialize the parameters into a SegmentSpecifierConfiguration. It just has to get the json string from this at line 89:
> 
> ```
>         String json = new Gson().toJson(configuration.getParameters());
> ```
> 
> So some extensions can optionally validate and deserialize the json string, when it's a complex object, and others can just use the map of parameters when it's a simple few values.

I think you are right. We should pass a Map<String, Object> from the server towards ITmfDataProviderSource.createDataProviderDescriptors() (and ITmfConfigurationSource.create()) and not use string. This would make it consistent to other TSP endpoints, avoids multiple serialization/deserialization steps, avoids having to deprecate, manage 2 APIs during deprecation period and manage removal of old API (in global). Overall, the server API will be simplier.

Extension can use map lookups and if wanted convert it to a JSON string and feed it to a POJO and custom deserializer, as showed in your example. using Gson.

@abhinava-ericsson FYI, the API will use default Map<String, Object> for JSON objects. The pull requests will be updated for that.